### PR TITLE
bpo-30068: add missing iter(self) in _io._IOBase.readlines when hint is present

### PR DESCRIPTION
**Pull request by zhangyangyu**

### CODE REVIEW

File: Modules/_io/iobase.c [hunk 1]
```diff
@@ -666,7 +666,13 @@ _io__IOBase_readlines_impl(PyObject *self, Py_ssize_t hint)
     }
 
     while (1) {
-        PyObject *line = PyIter_Next(self);
+        PyObject *it, *line;
+        it = PyObject_GetIter(self);
```
Comment (serhiy-storchaka) [inline Modules/_io/iobase.c]: This should be outside of the loop.

File: Misc/NEWS [hunk 2]
```diff
@@ -307,15 +307,17 @@ Extension Modules
 Library
 -------
 
+- bpo-30068: _io._IOBase.readlines will check if it's closed first when
+  hint is present.
+
+- bpo-29692: Fixed arbitrary unchaining of RuntimeError exceptions in
```
Comment (serhiy-storchaka) [inline Misc/NEWS]: Would be better not move this entry.

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]

File: Modules/_io/iobase.c (hunk 1 above)
Comment (zhangyangyu) [reply Modules/_io/iobase.c]: My foolish.

File: Modules/_io/iobase.c [hunk 3]
```diff
@@ -665,8 +665,14 @@ _io__IOBase_readlines_impl(PyObject *self, Py_ssize_t hint)
         return result;
     }
 
+    it = PyObject_GetIter(self);
```
Comment (serhiy-storchaka) [inline Modules/_io/iobase.c:667]: `it` is leaked.

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]

File: Modules/_io/iobase.c (hunk 3 above)
Comment (zhangyangyu) [reply Modules/_io/iobase.c:667]: Done

File: Modules/_io/iobase.c [hunk 4]
```diff
@@ -658,36 +658,43 @@ _io__IOBase_readlines_impl(PyObject *self, Py_ssize_t hint)
...
         }
         length += PyObject_Size(line);
         Py_DECREF(line);
 
         if (length > hint)
             break;
     }
+    Py_DECREF(it);
     return result;
+ error:
```
Comment (serhiy-storchaka) [inline Modules/_io/iobase.c]: Tiny nit: I prefer to add an empty line after `return` and before a label.

**Review by serhiy-storchaka — APPROVED** [review]

**Pull request merged**